adding Lexicase selector.

cliff-bohm opened the pull request.

cliff-bohm commented:
I have added docs here:
github.com/Hintzelab/MABE/wiki/Optimizer-Lexicase

cgnitash commented:
So LecicaseOpt doesn't generate max.csv. What's the reason?

cliff-bohm commented:
Max is meaningless in this context. Lexicase is a multi-objective optimizer so there is no obvious way to establish a meaningful max. We could allow the user to provide a max formula, but there is no need, all of the data to find this organism is available in snapshot files. btw, the same holds for Island optimizer.

There is another possible conversation about removing max.csv from archivist and making a feature of any optimizer where it makes sense, but that would require a significant change and I don't feel that this is sufficiently important to spend time on at the moment.

cgnitash commented:
Ok, makes sense. 

If it makes sense to get rid of max.csv at some point, let's make it an issue. That way we have a reminder of it when someone has time to work on it.

cgnitash commented:
The wiki says there are no sexes or sexual reproduction in LexiCaseOpt, but there's code that does stuff.

cgnitash commented:
Everything seems to work, i.e. the code compiles, and no segfaults were found for a small range of parameter values.

cliff-bohm commented:
a) I'm not sure I see the problem with the docs. the docs say there are no sexes or sexual selection which is correct.

b) If you want to do code simplifications you just as well do them now.

cgnitash commented:
a) Line 239 in the .cpp

b) Ok. Then I'll do the same for other PRs as well.

cgnitash commented:
I've made a PR #230  with edits for merging into this one. I thought git would auto-reference this, but apparently not.

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+
+	popFileColumns.clear();
+	if (recordOptimizeValues){
+		for (auto name : scoreNames) {
```
Should be auto &

cliff-bohm replied:
done

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+	std::vector<int> formulasOrder;
+	std::vector<int> sourceValues; // temp vector used to generate formulasOrder
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		sourceValues.push_back(i);
```
Should be std::iota

cliff-bohm replied:
done

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+			for (size_t i = 0; i < keepers.size(); i++) {
+				keeperScores.push_back(scores[keepers[i]][formulaIndex]);
+			}
+			std::sort(keeperScores.begin(), keeperScores.end());
```
This must be nth_element

cliff-bohm replied:
done

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+		}
+		// for each keeper, see if there are still a keeper, i.e. they have score >= scoreCutoff
+		for (size_t i = 0; i < keepers.size();) {
+			if (scores[keepers[i]][formulaIndex] >= scoreCutoff) {
```
This should be a std::remove_if

cliff-bohm replied:
this would require a redesign of the scores data structure since I need to remove from the outer vector based on the inner vector. a later refactoring should be considered.

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+		sourceValues.push_back(i);
+	}
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		int pick = Random::getIndex(sourceValues.size());
```
If Random:: provides its number generator, this could be a shuffle.

cliff-bohm replied:
done

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+			// get the range of scores
+			double maxScoreHere = scores[keepers[0]][formulaIndex];
+			double minScoreHere = maxScoreHere;
+			for (size_t i = 0; i < keepers.size(); i++) {
```
This should be std::min_element and std::max_element

cliff-bohm replied:
I think this would require the same type of restructuring as remove_if

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+	std::vector<double> aveScores;
+	std::vector<double> maxScores;
+	std::vector<double> minScores;
+	for (size_t fIndex = 0; fIndex < optimizeFormulasMTs.size(); fIndex++) {
```
push_backs should be avoided when the size and initial values of the vector are known up front.

cliff-bohm replied:
done
I am using:
'''	scores = std::vector<std::vector<double>>(population.size(), std::vector<double>(optimizeFormulasMTs.size(), 0));
'''
once per update to reset scores. This has to be done (or at least checked for) since population size could be variable.

cgnitash commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+		// for each org in population
+		killList.insert(population[i]); // add to kill list so that they are deleted in cleanup step
+		scores.push_back({});
+		for (size_t fIndex = 0; fIndex < optimizeFormulasMTs.size(); fIndex++) {
```
the min max ave stuff could be factored to use std::min_element std::max_element std::accumulate

cliff-bohm replied:
yes... I agree, but that would change score ... see if_remove above.

cgnitash commented:
Need to follow naming conventions.

cliff-bohm commented on Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp:
```diff
@@ -0,0 +1,268 @@
+			}
+			newPopulation.push_back(parents[0]->makeMutatedOffspringFromMany(parents));
+		}
+		nextPopulationSize++;
```
I see the point here. But I don't want to change it so that if in the future I want to add the option to produce more then one offspring per parent(s) selection I can add that more easily.

cliff-bohm commented:
I have addressed most issues (except var names and changes that would require non-trivial refactoring).

cgnitash commented:
Yeah, looks good. Have made some code transformations in #235 
This changes the results. Make sure that it's not qualitatively different if you want to merge, or that it's not a bug.

cgnitash commented:
The files don't include all the necessary headers. These must be added.

cgnitash commented:
Necessary changes are made in #235. There are also a number of code improvements that should probably be merged as well.

cliff-bohm commented:
This now contains most suggestions from #235, but also renamed the parameter that controls the epsilon method to epsilonRelativeTo (with values "rank" or "score"), fixed some type mismatch issues and cleaned up the code associated with epsilonRelativeTo score.

cliff-bohm commented:
Still waiting on a consensus for a naming standard for vars (which we will settle on Wednesday at worst.)
